### PR TITLE
[#285] Model properties change notification

### LINKED ISSUE #285: Add callback for properties update
We need to provide a functionality to allow users write code that will be executed on model property map update.

### LINKED ISSUE #390: Model container (pymodel.model.model) tries to de-serialize properties update callback
Model container (pymodel.model.model) tries to de-serialize properties update callback even if it uses for reading meta information.

```
[DYNAMIC MODEL XXX] Running shell script
+ cd .
+ legionctl build --docker-image-tag legion_ci_xxx-model_1.0-180829100718.11.47f77c5_2972 --push-to-registry localhost:31111/legion_model/xxx-model:1.0-180829100718.11.47f77c5 http://legion-core-nexus.default.svc.cluster.local:8081/repository/raw/xxx-model-1.0+180829100718.11.47f77c5.model
X-Legion-Model-Id:xxx-model
X-Legion-Model-Version:1.0
2018-08-29 10:11:26,538 - ERROR - No module named 'graph'
Traceback (most recent call last):
  File "/usr/local/bin/legionctl", line 150, in <module>
    args.func(args)
  File "/usr/local/lib/python3.6/dist-packages/legion/edi/deploy.py", line 59, in build_model
    container = legion.pymodel.Model.load(external_reader.path)
  File "/usr/local/lib/python3.6/dist-packages/legion/pymodel/model.py", line 272, in load
    instance._load_from_archive(path)
  File "/usr/local/lib/python3.6/dist-packages/legion/pymodel/model.py", line 248, in _load_from_archive
    self._on_property_update_callback = dill.load(callback_file)
  File "/usr/local/lib/python3.6/dist-packages/dill/dill.py", line 288, in load
    obj = pik.load()
  File "/usr/local/lib/python3.6/dist-packages/dill/dill.py", line 796, in _import_module
    return __import__(import_name)
ModuleNotFoundError: No module named 'graph'
```

### PR DESCRIPTION
**Pull request by kirillmakhonin**
This adds model callback that can be registered by model. In another thread we will run K8S watch that will monitor config map updates
This closes #285 and closes #390

### CODE REVIEW

File: legion/legion/model/__init__.py [hunk 1]
```diff
@@ -216,3 +216,17 @@ def define_property(name, initial_value):
         raise Exception('Context has not been defined')
 
     return _model.define_property(name, initial_value)
+
+
+def on_property_update(callback):
```
Comment (akharlamov) [inline legion/legion/model/__init__.py]: I'd make it on_property_change as we do not distinguish b/w prop create/delete/update here.

File: legion/legion/pymodel/model.py [hunk 2]
```diff
@@ -191,6 +195,9 @@ def __init__(self, model_id, model_version):
         self._endpoints = {}  # type: dict or None
         self._path = None  # type: str or None
 
+        self._properties_update_thread = None  # type: threading.Thread or None
```
Comment (akharlamov) [inline legion/legion/pymodel/model.py]: This should be inside K8SProperties class

File: legion/tests/test_deploy.py [hunk 3]
```diff
@@ -73,6 +73,7 @@ def test_summation_model_build_and_query(self):
                 'invalid batch invocation result'
             )
 
+    @unittest2.skip
```
Comment (akharlamov) [inline legion/tests/test_deploy.py]: Why?

**Comment by kirillmakhonin** [comment]
@akharlamov?

**Comment by kirillmakhonin** [comment]
test this please

**Comment by kirillmakhonin** [comment]
@akharlamov @dsuslov @aliaksandr-d @alexander-semenets please review

File: deploy/helms/legion/templates/edi/edi-rbac.yaml [hunk 4]
```diff
@@ -10,7 +10,7 @@ rules:
   verbs: ["watch", "list", "create", "delete"]
 - apiGroups: [""] # core API group
   resources: ["configmaps", "secrets"]
-  verbs: ["list", "create", "get", "patch", "update"]
+  verbs: ["list", "create", "get", "patch", "update", "list"]
```
Comment (aliaksandr-d) [inline deploy/helms/legion/templates/edi/edi-rbac.yaml]: 'list' verb has been already added.

File: legion_test/legion_test/robot/model.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
+#
+#    Copyright 2017 EPAM Systems
```
Comment (aliaksandr-d) [inline legion_test/legion_test/robot/model.py]: Please, update to 2018y

**Comment by alexander-semenets** [comment]
despite of the mentioned issues looks good

File: legion/legion/serving/pyserve.py [hunk 6]
```diff
@@ -216,19 +216,24 @@ def init_model(application):
...
+        id(model_container.properties)
+    ))
     legion.model.set_properties(model_container.properties)
 
     # Force reload if code run in a cluster and model required any properties
     if legion.k8s.utils.is_code_run_in_cluster() and model_container.required_props:
         legion.model.properties.load()
-
-    return model_container
+        legion.model.properties.start_update_watcher()
```
Comment (kirillmakhonin) [inline legion/legion/serving/pyserve.py]: Add logs why failed to start

File: tests/python/test_k8s_properties_storages.py [hunk 7]
```diff
@@ -95,6 +95,84 @@ def test_overwrite_config_map_storage(self):
...
+        storage_to_read = legion.k8s.K8SSecretStorage.retrive(storage_name, TEST_ENCLAVE_NAME)
+
+        def listener():
+            for event, new_data in storage_to_read.watch():
+                LOGGER.info('Got new event: type={}, new_data={}'.format(event, new_data))
+                events.append((event, new_data))
+
+        with legion_test.utils.ContextThread(listener):
+            LOGGER.debug('Waiting before updating')
+            time.sleep(5)
```
Comment (kirillmakhonin) [inline tests/python/test_k8s_properties_storages.py]: Change sleep to iterated sleep

File: tests/robot/2_check_models.robot [hunk 8]
```diff
@@ -13,6 +17,26 @@ Running, waiting and checks jobs in Jenkins
...
+    ${model_id}    ${model_version} =   Test model pipeline  ${MODEL_WITH_PROPS}  ${MODEL_TEST_ENCLAVE}
+    Log                                 Model with id = ${model_id} and version = ${model_version} has been deployed
+    ${edge}=        Build enclave EDGE URL  ${MODEL_TEST_ENCLAVE}
+
+    Log             Updating property to start value and invoking model with check
+    Update model property key  ${MODEL_TEST_ENCLAVE}  ${model_id}  ${model_version}  ${MODEL_WITH_PROPS_PROP}  1
+    ${response}=    Invoke model API  ${model_id}  ${model_version}  ${edge}  ${MODEL_WITH_PROPS_ENDPOINT}  a=1  b=2
+    Should Be Equal As Integers     ${response['result']}  30
+
+    Sleep  10s
```
Comment (kirillmakhonin) [inline tests/robot/2_check_models.robot]: Remove sleep, add validation of current model properties

**Comment by kirillmakhonin** [comment]
TODOs:

- Add model properties introspection endpoint
- Add flag to `legionctl undeploy` to remove existed model properties
- Add tests for default properties creation process
- Add tests for ignoring overwriting of model properties
- Add negative tests for model properties

File: legion/legion/serving/pyserve.py (hunk 6 above)
Comment (kirillmakhonin) [reply legion/legion/serving/pyserve.py]: done

**Comment by kirillmakhonin** [comment]
ok to test

**Comment by kirillmakhonin** [comment]
ok to test

**Comment by kirillmakhonin** [comment]
ok to test

**Comment by kirillmakhonin** [comment]
ok to test

**Comment by kirillmakhonin** [comment]
ok to test

**Comment by kirillmakhonin** [comment]
ok to test

File: legion/tests/test_store.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    @classmethod
+    def setUpClass(cls):
+        """
+        Enable logs on tests start, setup connection context
+
+        :return: None
+        """
+        logging.basicConfig(level=logging.DEBUG)
+
+    @unittest2.skip
```
Comment (dsuslov) [inline legion/tests/test_store.py]: enable it

File: legion/legion/templates/Dockerfile.tmpl [hunk 10]
```diff
@@ -39,11 +39,7 @@ CMD ["/usr/bin/supervisord", "-c", "/etc/supervisor/conf.d/supervisord.conf" ]
 WORKDIR /app
 
 RUN {{CREATE_SYMLINK_COMMAND|safe}}
-
```
Comment (dsuslov) [inline legion/legion/templates/Dockerfile.tmpl]: please, remove this line

**Comment by kirillmakhonin** [comment]
ok to test

**Comment by kirillmakhonin** [comment]
ok to test

**Comment by kirillmakhonin** [comment]
test this please

**Comment by kirillmakhonin** [comment]
test this please

**Pull request merged**